plat/common/include/x86/paging.h

```diff
@@ -249,7 +249,7 @@ static __paddr_t x86_pg_maxphysaddr;
 int ukarch_paddr_range_isvalid(__paddr_t start, __paddr_t end)
 {
 	UK_ASSERT(start <= end);
-	return (X86_PG_VALID_PADDR(start) && X86_PG_VALID_PADDR(end));
+	return (X86_PG_VALID_PADDR(start) && X86_PG_VALID_PADDR(end - 1));
```

Comment:
Hmm, wait does this mean that this change should be done for ARM as well? This looks inconsistent. Am I missing something?

Reply:
Hey, I think this solution is not right after all. It is the user of the function that uses it incorrectly by overestimating the end address, so it should be fixed there instead. What do you think about removing this commit and keeping the previous one?

Reply:
> I think this solution is not right after all

Agreed.

> It is the user of the function that uses it incorrectly by overestimating the end address

Where was this wrong use again?

> What do you think about removing this commit and keeping the previous one?

I agree :+1:. This would also align this with ARM, right? So this means we'll only keep the first commit here.

Reply:
> Where was this wrong use again?

It was a case where UEFI contained an address range that would trigger an assertion failure. There are some changes to my environment, so I could not get this behavior again :smiling_face_with_tear:. The bug was triggered when  `ukplat_pt_add_mem` was called on the memory range.

 ```
[    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  0000feffc000-0000ff000000 000000004000 r-- 00000000feffc000 rsvd
[    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  0000ffc00000-000100000000 000000400000 rw- 00000000ffc00000 rsvd
[    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  00fd00000000-010000000000 000300000000 r-- 000000fd00000000 rsvd
```

Now, looking at it again, 010000000000 would be out of valid address space by 1, where the max address is supposed to be 0xFFFFF.... This is still true if we align the max physical address definition to ARM64.

Reply:
> > Where was this wrong use again?
> 
> It was a case where UEFI contained an address range that would trigger an assertion failure. There are some changes to my environment, so I could not get this behavior again 🥲. The bug was triggered when `ukplat_pt_add_mem` was called on the memory range.
> 
> ```
> [    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  0000feffc000-0000ff000000 000000004000 r-- 00000000feffc000 rsvd
> [    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  0000ffc00000-000100000000 000000400000 rw- 00000000ffc00000 rsvd
> [    0.000000] Info: [libkvmplat] <bootinfo.c @  101>  00fd00000000-010000000000 000300000000 r-- 000000fd00000000 rsvd
> ```
> 
> Now, looking at it again, 010000000000 would be out of valid address space by 1, where the max address is supposed to be 0xFFFFF.... This is still true if we align the max physical address definition to ARM64.

Ah, I remember now... this means this stuff can also happen for ARM with such a region. What if we did it like this for this PR
```diff
--- a/arch/arm/arm64/include/uk/asm/paging.h
+++ b/arch/arm/arm64/include/uk/asm/paging.h
@@ -146,8 +146,8 @@ struct ukarch_pagetable {
 /* Any PTE with bit[0] == 0 is invalid */
 #define PT_Lx_PTE_INVALID(lvl)         0x0UL
 
-#define ARM64_PADDR_MAX                        ((1ULL << ARM64_PADDR_BITS) - 1)
-#define ARM64_VADDR_MAX                        ((1ULL << ARM64_VADDR_BITS) - 1)
+#define ARM64_PADDR_MAX                        (1ULL << ARM64_PADDR_BITS)
+#define ARM64_VADDR_MAX                        (1ULL << ARM64_VADDR_BITS)
 
 #define ARM64_PADDR_VALID(paddr)       (paddr <= (__paddr_t)ARM64_PADDR_MAX)
 #define ARM64_VADDR_VALID(vaddr)                               \
--- a/include/uk/arch/paging.h
+++ b/include/uk/arch/paging.h
@@ -370,7 +370,8 @@ int PAGE_Lx_IS(__pte_t pte, unsigned int lvl);
  * 64 bits for the virtual address.
  *
  * @param start the start of the virtual address range
- * @param end the end of the virtual address range
+ * @param end the end of the virtual address range, where end is equal to
+ *       (start + length of range)
  *
  * @return a non-zero value if the addresses in the range are supported
  */
@@ -386,7 +387,8 @@ int ukarch_vaddr_range_isvalid(__vaddr_t start, __vaddr_t end);
  * installed in the system.
  *
  * @param start the start of the physical address range
- * @param end the end of the physical address range
+ * @param end the end of the physical address range, where end is equal to
+ *       (start + length of range)
  *
  * @return a non-zero value if the addresses in the range are supported
  */
 --- a/plat/common/include/x86/paging.h
+++ b/plat/common/include/x86/paging.h
@@ -244,7 +244,7 @@ pgarch_kunmap(struct uk_pagetable *pt __unused, __vaddr_t vaddr __unused,
 
 static __paddr_t x86_pg_maxphysaddr;
 
-#define X86_PG_VALID_PADDR(paddr)      ((paddr) < x86_pg_maxphysaddr)
+#define X86_PG_VALID_PADDR(paddr)      ((paddr) <= x86_pg_maxphysaddr)
 
 int ukarch_paddr_range_isvalid(__paddr_t start, __paddr_t end)
 {
```

I think this way the two arch's would be aligned and the addition to the function's description could clarify to its potential user what it expects. And have this all in a single commit like:
`arch/*: Align paddr/vaddr range validation between x86 and ARM64`

Hmm, the validation of the `vaddr` is also misaligned: x86 only checks for cannonical addresses and thus no need for range, while ARM checks for range. Commit `b0c5224663c0a3b133703012892bff11d3e264a0` seems to be clear as to why: the upper canonnical range is considered as invalid so I guess it is fine.

```
    arch/arm: plat/common: Implement paging support for arm64
    
    This commit implements virtual memory support for arm64 as defined
    by VMSAv8-64. The implementation supports 48-bit VAs with a 4KiB granule,
    and uses the lower address space controlled by TTBR0_EL1, i.e.
    0x0000_0000_0000_0000 - 0x0000_ffff_ffff_ffff.
    
    The last 512GiB (0x0000_ff80_0000_0000) are reserved, and are direct-mapped
    to the first 512GiB (0x0000_0000_0000_0000) to allow fast translation of
    the pagetables.
    
    The upper range controlled by TTBR1_EL1 at 0xffff_0000_0000_0000 is
    configured as invalid.
 ```

Since you were part of the discussion of this commit @michpappas @razvanvirtan. What do you whink about the proposed changes.

I would go with the proposed diff above. What do you think? Could there be any potential pitfalls? :thinking:

Reply:
Also, sidenote, design-wise, it would have probable made more sense if `ukarch_*_range_isvalid` simply took a start address and the length of the range, instead of start/end. But, oh well, maybe it's a discussion for some other time.

Reply:
I agree with the fact that taking the address and length as parameters in `ukarch_paddr_range_isvalid()` makes more sense for the future.

For the purpose of this PR, I think we should also keep the logic for the macros names. This means that on both ARM64 and x86, the max paddr macros should be defined as `(1 << PADDRBITS) - 1`, because they are suppose to represent the maximum usable physical address, not the physical memory size. Therefore, I suggest the following changes:

- we keep the first commit in this PR, so that we correctly define `x86_pg_maxphysaddr`
- we keep the second commit in this PR, correctly defining `ukarch_paddr_range_isvalid()` on x86
- we add a similar fix for `ukarch_paddr_range_isvalid()` on ARM64 (checking for `end - 1` instead of `end`)
- we add the comments suggested above by @mogasergiu, so that we have the `ukarch_paddr_range_isvalid()` function clearly documented

WDYT?

Reply:
Yes, I also like that. Actually, more than what I have previously suggested. You are making a very good point @razvanvirtan. :brain: :monocle_face: :1st_place_medal:

Reply:
Sounds good to me :100:.

Reply:
To me it seems that the root of the problem is that the current interface is easy to misuse, so instead of working around the problem, perhaps we should apply @mogasergiu's suggestion to pass (`addr, `len`)? My guess would be that it should require minimal effort. But since folks already came into consensus, the currently proposed approach is also okay.

Reply:
```
include/uk/arch/paging.h:380:	ukarch_vaddr_range_isvalid(vaddr, vaddr)			\
include/uk/arch/paging.h:396:	ukarch_paddr_range_isvalid(paddr, paddr)
lib/ukfallocbuddy/fallocbuddy.c:498:	UK_ASSERT(ukarch_paddr_range_isvalid(start, start + len));
lib/ukfallocbuddy/fallocbuddy.c:1369:	UK_ASSERT(ukarch_paddr_range_isvalid(paddr, paddr + len));
lib/ukfallocbuddy/fallocbuddy.c:1443:	UK_ASSERT(ukarch_paddr_range_isvalid(paddr, paddr + len));
lib/ukvmem/vma_dma.c:41:	UK_ASSERT(ukarch_paddr_range_isvalid(args->paddr, args->paddr + len));
lib/ukvmem/vmem.c:65:	UK_ASSERT(ukarch_vaddr_isvalid(vas->vma_base));
lib/ukvmem/vmem.c:218:	UK_ASSERT(ukarch_vaddr_range_isvalid(vstart, vend));
lib/ukvmem/vmem.c:665:	UK_ASSERT(ukarch_vaddr_range_isvalid(va, va + len));
plat/common/include/arm/arm64/paging.h:349:	UK_ASSERT(ukarch_paddr_range_isvalid(start, start + len));
plat/common/include/x86/paging.h:326:	UK_ASSERT(ukarch_paddr_range_isvalid(start, start + len));
plat/common/paging.c:270:	UK_ASSERT(ukarch_paddr_range_isvalid(start, start + len));
plat/common/paging.c:310:	UK_ASSERT(ukarch_paddr_range_isvalid(start, start + len));
plat/common/paging.c:387:	UK_ASSERT(ukarch_vaddr_isvalid(vaddr));
plat/common/paging.c:536:	UK_ASSERT(ukarch_vaddr_range_isvalid(vaddr, vaddr + len));
plat/common/paging.c:541:		UK_ASSERT(ukarch_paddr_range_isvalid(paddr, paddr + len));
plat/common/paging.c:959:		UK_ASSERT(ukarch_vaddr_range_isvalid(vaddr, vaddr + len));
plat/common/paging.c:1240:		UK_ASSERT(ukarch_vaddr_range_isvalid(vaddr, vaddr + len));
```

These would be the places that would need necessary changes if the interface were to be changed. This might require extensive testing :thinking:.

> My guess would be that it should require minimal effort. But since folks already came into consensus, the currently proposed approach is also okay.

:+1: If @kha-dinh agrees to also changing those bits and pieces of code shown above (hopefully I did not miss any) as well as the interface signature, I am fine with that. But I'd say let's go with what @razvanvirtan suggested for now, and put this change somewhere in the backlog if @kha-dinh decides to not do these changes.

Feel free to go with whatever you like @kha-dinh :pray:. Either way is fine IMO :+1:.